### PR TITLE
Move core service from testbot to client

### PR DESCRIPTION
**Pull request by rcooke-warwick**
This change moves the core container off of the Worker, onto the client. This is towards merging core and client - having tests execute remotely and have a simple interface for the worker. 

Big changes at a glance:
- worker endpoint for start - to reserve the worker for a given test run. Includes heartbeat with client in case of connection loss
- Commands are executed on the DUT by SSH'ing into the worker via the balena proxy, and then SSH'ing from there into the DUT. The test suite must be updated to register an ssh key with the cloud so these commands are authenticated
- Tunnels can be made from the client/core to the DUT. This is useful for communicating with the DUT supervisor from tests, and also for doing "local" pushes of containers to the DUT. This is done by establishing a tunnel from the worker to the DUT, and then from the core to the worker. 
- HDMI capture now sends the captured images to the core upon request
- Serial logs can be fetched from the worker directly

Todo:

- [x]  tests must be updated to work with this setup. In progress atm
- [x]  Hup suite
- [x]  test cloud suite - but in theory should work
- [ ] e2e test suite

Resolves: https://github.com/balena-os/leviathan/issues/651
Change-type: major
Signed-off-by: Ryan Cooke <ryan@balena.io>

### CODE REVIEW

**Review by jakogut — COMMENTED** [review]
Much simpler, good progress.

File: worker/lib/index.ts
Comment (vipulgupta2048) [reply worker/lib/index.ts:2]: Ryan, try the new npm hooks we got for lining and fixing code. `npm run lint:fix`

**Comment by rcooke-warwick** [comment]
@jakogut @vipulgupta2048 @klutchell I've used an npm ssh package where I can. 

I've also tried to take into account the scenario where there is a virtualised worker (in this case the core can SSH directly to the DUT, and local push directly to the DUT) and a local-mode testbot worker (in this case you don't access the worker through the VPN). The resulting code is quite messy imo - a lot of if-else statements. Any ideas of how to clean that up are appreciated.

File: core/lib/common/utils.js
Comment (jakogut) [reply core/lib/common/utils.js]: @rcooke-warwick After reviewing several SSH packages, I'm partial to ssh2-promise over ssh2 or node-ssh.

Ssh2 uses a callback interface, and although it supports the functionality we need, it's lower level than I'd like.

Node-ssh wraps ssh2 in a Promise interface, and is nicer to use, but doesn't support automatic reconnection. When a connection drops, you're back to using a callback interface to handle the error or closed connection, and reconnecting isn't straightforward.

ssh2-promise has the nicer Promise-based interface of node-ssh, with automatic reconnection, connection chaining, SFTP support, etc.

File: core/lib/common/utils.js
Comment (jakogut) [inline core/lib/common/utils.js:39]: This creates a new SSH connection per command, which is unnecessary overhead. We should be able to open a connection and multiplex commands through it until we're done, or the connection is broken and we need to reconnect. This won't make a big difference locally, but could save a lot of time when the round-trip latency is high.

Also, I understand that this is a copy of how SSH connections are currently done, so it's not worse than the current approach, but I wanted to point it out.

File: core/lib/common/worker.js
Comment (jakogut) [inline core/lib/common/worker.js:346]: @rcooke-warwick This might be easier with ssh2-promise https://github.com/sanketbajoria/ssh2-promise#connect-to-ssh-server-via-hopping

**Review by jakogut — COMMENTED** [review]
I'm imagining it would be simpler to move SSH connection handling to the suite(s), as the client can connect directly to the testbot, and indirectly to the test device through the testbot, with no need for the core service as an intermediate. This will of course require test modifications.

I wonder if we could differentiate between a direct connection to the test device, such as when running virtualized, and a proxied connection, such as when connecting through a testbot, by setting an intermediate SSH config?

File: core/lib/common/worker.js
Comment (rcooke-warwick) [reply core/lib/common/worker.js:346]: @jakogut the only problem with the server hopping commands with the ssh2 library (as far as I can tell), is that it doesn't account for the `host <UUID>` option thats required to get into a device connected to balena cloud via the proxy.

**Comment by rcooke-warwick** [comment]
@jakogut this commit I'm about to push cleans things up also - it creates 3 ssh tunnels (only if a testbot is being used). If its a qemu test, this isn't needed and the DUT can be ssh'd and local pushed to directly.

- port 43434 to the DUT supervisor 
- port 1375 to the DUT engine
- port 22222 to the DUT openssh server

In the case of a testbot being used to drive the DUT, the DUT can be ssh'd into "directly" using `localhost:22222`, using a node ssh library like ssh2. Using a node library. (In my commit I've stuck to the one being used at the moment, node-ssh).

I've used the worker url given to the test as part of the config to determine whether or not its a virtualised or testbot scenario.

**Comment by rcooke-warwick** [comment]
Once this PR is merged I think we can make another PR to merge core and client. I imagine that it would be a case of spawning `suite` subprocesses from the client container, and executing the tests in there, instead of having the client speak to the intermediate core container.

**Comment by rcooke-warwick** [comment]
This PR still also doesn't include a solution for the HUP suite. To carry out the HUP suite, we agreed we would try to fix the script that allows you to HUP from a file locally - until that is done, then this version of leviathan won't be able to do the hup suite

**Comment by jakogut** [comment]
> Once this PR is merged I think we can make another PR to merge core and client.

@rcooke-warwick Agreed, that's a good plan.

**Comment by klutchell** [comment]
> This PR still also doesn't include a solution for the HUP suite. To carry out the HUP suite, we agreed we would try to fix the script that allows you to HUP from a file locally - until that is done, then this version of leviathan won't be able to do the hup suite

I don't want to block this PR any further, and the existing HUP suite is pretty broken anyway. Why don't we disable the HUP suite at the jenkins or meta-balena level for now to unblock everything else? That way we are testing against the new leviathan architecture going forward.

Also, do we have a PR in meta-balena for the suite changes required with this PR? They need to be merged around the same time obviously.

**Comment by rcooke-warwick** [comment]
Will clean this up so we can merge asap. Also will make the corresponding meta balena PR

File: core/Dockerfile.template
Comment (vipulgupta2048) [inline core/Dockerfile.template:20]: Might need to rebase this, we already did this

File: core/lib/common/worker.js
Comment (vipulgupta2048) [reply core/lib/common/worker.js:336]: @rcooke-warwick Do run `npm run lint:fix` to possibly resolve these.

**Review by vipulgupta2048 — COMMENTED** [review]
Do update worker's package-lock

File: Makefile
Comment (klutchell) [inline Makefile:84]: I plan to perform some cleanup of this Makefile after the merge, some of the existing functionality is bugging me, like lack of parallel builds with docker-compose.

**Comment by rcooke-warwick** [comment]
@klutchell just cleaning up formatting and will squash the commits. I've tested the OS and cloud suite with the testbot worker, and with the qemu worker and all tests pass (both production and development mode images) (given some minor changes to the `os` test suite) - After the cleanup I'll run the tests one more time on each worker to be sure, and then remove the draft tag.

I think while there may be some things that aren't optimal in this PR - we should probably consider getting it through the gate, and make further improvements iterative. There's already a bunch of big changes here and I think further removal of endpoints in the worker etc will just delay things further. Once this change is merged, further improvements become a lot quicker and easier as most of the logic sits outside the app on the hardware. 

As for the hup suite, this PR adds a function to send a file/folder to the DUT using rsync over ssh - we can use that to send the image to the DUT and when the hup script is fixed, hup from the local file we just transferred

**Comment by rcooke-warwick** [comment]
@balena-ci rebase

File: docker-compose.yml
Comment (klutchell) [inline docker-compose.yml:17]: Can we also remove this service from `docker-compose.local.yml`?

File: docker-compose.client.yml
Comment (klutchell) [inline docker-compose.client.yml:23]:
```suggestion
```
Do we need/use this value anymore?

File: client/lib/index.js
Comment (klutchell) [inline client/lib/index.js:300]: nitpick, but in a future PR we should make this configurable

File: docker-compose.client.yml
Comment (klutchell) [reply docker-compose.client.yml:23]: Looks like we have some stale values and defaults in here:
https://github.com/balena-os/leviathan/blob/ryan/core-to-client/core/config/default.js

File: core/lib/common/worker.js
Comment (jakogut) [inline core/lib/common/worker.js:436]: Are these functions meant to be internal, or used from test suites as well? If they're internal only, I think they're fine. If they're meant to be used from suites, I see a number of problems.

Comment (rcooke-warwick) [reply core/lib/common/worker.js:436]: At the moment, I'm only using the sendFileToWorker internally.

The send file to DUT I'm using in the HUP suite:  https://github.com/balena-os/meta-balena/blob/ryan/tests-update-for-leviathan/tests/suites/hup/suite.js#L91

File: client/lib/index.js
Comment (rcooke-warwick) [reply client/lib/index.js:300]: agreed

File: core/lib/common/worker.js
Comment (jakogut) [reply core/lib/common/worker.js:435]: @rcooke-warwick If these functions are for internal use only, they can be easily changed, and my feedback is moot. If they're used by the test suites, they're an interface, and changing them will be breaking, so we'll want to get it right the first time, if possible.

Rsync is an implementation detail, I don't see any value to including it in the function name. If anything, it'll confuse things in the future if we change the implementation.

In fact, worker/DUT also appears at first glance to be a parameter, I don't currently understand the need for separate functions here. At the end of the day, we're transferring files from one host to another. Which host does not need to be encoded in the function name.

**Comment by rcooke-warwick** [comment]
@klutchell, with latest commits to this and the meta balena PR, I have all 3 suites passing on both testbot + qemu

**Comment by rcooke-warwick** [comment]
I have had to revert: https://github.com/balena-os/leviathan/pull/624 - as to preload the image, we need to find its path from the hsotOS, as we're using the hosts docker daemon to do the preloading. AFAIK, to do that we need the image to sit in a volume.

**Comment by klutchell** [comment]
> I have had to revert: #624 - as to preload the image, we need to find its path from the hsotOS, as we're using the hosts docker daemon to do the preloading. AFAIK, to do that we need the image to sit in a volume.

Yeah, I found the same thing actually, we do need the image to be on a volume to make it easier to find on the host.

**Comment by rcooke-warwick** [comment]
@klutchell that change also had the added problem of us not clearing that `/tmp` directory at the end of every test run, so unpacked images get stockpiled

**Comment by rcooke-warwick** [comment]
@balena-ci rebase

**Comment by rcooke-warwick** [comment]
@klutchell @vipulgupta2048 , thoughts on this PR? All suites pass for me on testbot + qemu

**Comment by klutchell** [comment]
@rcooke-warwick Let's merge it and retest the changes to the suites. Worst case, we can pin Jenkins leviathan-genericx86-64 to an older release via the branch field.

Note that we are NOT bumping the testbot fleet today as we need balena-raspberrypi to pick up the new suites.

**Comment by klutchell** [comment]
One more test I want to run, on Jenkins, hang tight.

**Comment by klutchell** [comment]
@rcooke-warwick I think we can't merge this until we have solved the cloud suite failing to connect to the dashboard.

If we merge this we will be blocked on merging the corresponding suite changes until we get lucky on a build of the new suites, as the issue is still present last time I tested both of these branches in Jenkins.

Also, I would like to add some code to the updated suites to indicate which leviathan branch (or version) they are using. We can make some assumptions in the Jenkins job that if we do not provide a leviathan branch/version we should assume the version prior to this merge.

I hope to have something soon, but it might just be a text file that is read by the yocto-device job and sets the leviathanBranch job parameter.

**Comment by klutchell** [comment]
@rcooke-warwick I am seeing good results with qemu but I've seen this a few times when testing on the `balena/testbot-personal` rig with my testbot running this release.
```
client_1  | [2022-02-25T23:03:38.698Z][8f854c3-cloud] Preparing to flash

client_1  | [2022-02-25T23:05:56.345Z][8f854c3-cloud] # Subtest: Managed BalenaOS release suite
client_1  | [2022-02-25T23:05:56.346Z][8f854c3-cloud]     not ok 1 - Unexpected end of TCP connection
```

**Comment by rcooke-warwick** [comment]
hey @klutchell I haven't been able to reproduce the above 2 errors locally. 

As for the first, I will add a retry to the flashing mechanism in case the VPN flakes mid flash
As for the second, that error signals that post flashing, your DUT never came online. I haven't been able to repdouce that, what image were you testing with?

**Comment by klutchell** [comment]
> hey @klutchell I haven't been able to reproduce the above 2 errors locally.
> 
> As for the first, I will add a retry to the flashing mechanism in case the VPN flakes mid flash As for the second, that error signals that post flashing, your DUT never came online. I haven't been able to repdouce that, what image were you testing with?

Thanks for checking! I will revisit these tests after we get some other things out of the way. In theory we can merge this after addressing Joseph's comments because we will be able to retest with the rig when bumping meta-balena in balena-raspberrypi.

File: core/lib/common/worker.js
Comment (jakogut) [inline core/lib/common/worker.js:251]: Nitpick:
- No point having a catch handler that rethrows the error here, especially when the promise is awaited immediately after. Just return the promise. Not a blocking suggestion, just a style/readability thing.

As an example, instead of doing:
```JavaScript
async foo() {
  const bar = baz().catch(e => {throw e});
  await bar;
}
```

You can do instead:
```JavaScript
async foo() {
    return baz();
}

**Review by jakogut — APPROVED** [review]
I left some nitpicks, none of them blocking. Looks fine to me.

**Pull request merged**
